app/src/main/java/de/westnordost/streetcomplete/data/meta/OsmTaggings.kt

```diff
@@ -32,32 +32,35 @@ val MAXSPEED_TYPE_KEYS = setOf(
-    // due to the multitude of these, let's keep the whole ref:.* space
-    "ref:.*",
-    "source",
-    // source tags referring to building / address / geometry (from cursory look at taginfo on "source:")
-    "source:(${listOf(
-        "date",
-        // referring to address / ref...
-        "addr", "address", "housenumber", "street", "conscriptionnumber", "streetnumber", "postcode", "postal_code", "ref",
-        // referring to geometry
-        "geometry", "position", "shape", "outline", "location",
-        // referring to building
-        "building", "heritage", "height", "architect", "ele"
-    ).joinToString("|")})(:.*)?",
+val KEYS_THAT_SHOULD_BE_REMOVED_WHEN_SHOP_IS_REPLACED = listOf(
+    "shop", "craft", "amenity", "shop_1", "old_shop", "craft_1",
+    // obsoleted information
+    "disused:.*", "was:.*", "damage", "source:damage", "created_by", "check_date",
+    // classifications / links to external databases
+    "fhrs:.*", "fvst:.*", "ncat", "nat_ref", "gnis:.*", "winkelnummer",
+    // names and identifications
+    "name(:.*)?", "name_1", ".*_name(:.*)?", "noname", "branch(:.*)?", "brand(:.*)?", "network", "operator(:.*)?", "ref", "ref:vatin", "designation",
+    // contacts
+    "contact:.*", "website", "facebook", "email", "phone(:.*)?", "phone_1", "mobile", "fax", "url", "source_ref:url", "owner",
+    // payments
+    "payment:.*", "currency:.*", "cash_withdrawal(:.*)?", "fee",
+    // generic shop/craft attributes
+    "opening_hours(:.*)?", "check_date:opening_hours", "internet_access(:.*)?", "second_hand", "self_service", "license:.*", "bulk_purchase", ".*:covid19", "description(:.*)?",
+    "min_age", "supermarket(:.*)?", "operational_status", "trade", "wholesale", "sale", "smoking", "origin", "attraction", "strapline", "dog",
+    // food and drink details
+    "bar", "cafe", "microbrewery", "brewery", "real_ale", "distillery", "drink:.*", "cocktails", "alcohol", "wine:region", "wine_region", "diet:.*", "cuisine",
+    "organic", "produced_on_site", "food", "pastry", "product", "produce", "chocolate", "fair_trade", "butcher", "reservation", "takeaway(:.*)?", "delivery", "caterer",
+    // fuel stations and boatbuilders
+    "fuel:.*", "compressed_air", "hgv:lanes", "seamark:.*", "waterway",
+    // related to repair shops/crafts
+    "service(:.*)?", "motorcycle:.*", "repair", ".*:repair", "electronics_repair(:.*)?",
+    // hairdresser
+    "unisex", "male", "female",
+    // accomodation & layout
+    "rooms", "stars", "accommodation", "beds", "capacity(:persons)?", "tourism", "laundry_service",
+    // misc specific attributes
+    "clothes", "beauty", "tobacco", "carpenter", "furniture", "lottery", "sport", "leisure", "dispensing", "tailor:.*",
+    "books", "agrarian", "musical_instrument", "massage", "parts", "post_office(:.*)?", "religion", "rental", "tickets:.*", "goods_supply",
 ).map { it.toRegex() }
```

Comment:
add
- wifi
- instagram, other contact-stuff found in contact:.?
- hairdresser, shoes, butcher, electronics_repair, pet, ... and other sub-tags of shops/crafts/amenity that are named like their shop/craft/amenity value.

Also, `isKindOfShopExpression` also includes some `leisure`, some `office` and one `tourism`. So, these need to be removed as well plus their properties.

Reply:
>     * wifi
>     * instagram, other contact-stuff found in contact:.?

added.

>     * hairdresser, shoes, butcher, electronics_repair, pet, ... and other sub-tags of shops/crafts/amenity that are named like their shop/craft/amenity value.

I'm not sure I follow. Do you mean eg. `shop=butcher` should also remove `butcher=*` ? I already do that (at least if that subkey was used at least `0.01%` or more times (or if it was mentioned in `id-tagging-schema/data/presets/`). Some do not exist, though (like `hairdresser=*`). Or did I miss some? 

> Also, `isKindOfShopExpression` also includes some `leisure`, some `office` and one `tourism`. So, these need to be removed as well plus their properties.

I've added them as main keys now (their properties were already added). I didn't initially as I (it seems wrongly) assumed main tag (the one the quest is about) is removed automatically, and that those were just extra keys to be removed beside that main tag. (That would be much nicer!)

Removing all `office=*` when `shop=convenience` will work (as in current PR), but is not ideal - for example sometimes there is a building which will be tagged with both `shop=*` and `office=*`, which are unrelated, and marking a shop vacant would remove information about offices (which it should not do). Yes, technically, it is a problem when **_tagging anything but the building_** on building closed-way, but is widely used practice unfortunately.

- Ideally, I would remove that main tags (`amenity`, `shop`, `craft`, ...) from `KEYS_THAT_SHOULD_BE_REMOVED_WHEN_SHOP_IS_REPLACED`, and the kotlin function would remove main tag that triggered the quest separately (in addition to that list). That would be preferable. (I wouldn't know how to do it myself, even if it is possible, unfortunately).
- if that is not possible, can one use `key=value` instead of just `key` in `KEYS_THAT_SHOULD_BE_REMOVED_WHEN_SHOP_IS_REPLACED` ? So I could say (in order to minimize damage) to remove just `office=insurance`, `office=travel_agent`, `office=tax_advisor` etc. instead of all `office=*` ?

Reply:
>  (it seems wrongly) assumed main tag (the one the quest is about) is removed automatically,

This might be a good candidate for a test.

Reply:
> I'm not sure I follow. Do you mean eg. shop=butcher should also remove butcher=* ?

Yes. Looks like I overlooked that. But I also meant generally. Try searching in taginfo for keys named like shop/leisure/craft/... values. I listeed some I found, but I did not look thoroughly.

> I've added them as main keys now (their properties were already added). I didn't initially as I (it seems wrongly) assumed main tag (the one the quest is about) is removed automatically

Well, right now, **everything** except the specified keys is removed. So when you want to turn this around, you also need to mention every tag that can be removed (i.e. is in that "is kind of shop" expression)... maybe worth a comment in the code, that if the "is kind of shop" expression is extended, your whitelist should be checked too.

Reply:
> Ideally, I would remove that main tags (amenity, shop, craft, ...) from KEYS_THAT_SHOULD_BE_REMOVED_WHEN_SHOP_IS_REPLACED, and the kotlin function would remove main tag that triggered the quest separately

Not possible. the filter expression just returns true or false (triggered or not), it doesn't say exactly which part of the expression matched.

> if that is not possible, can one use key=value instead of just key in KEYS_THAT_SHOULD_BE_REMOVED_WHEN_SHOP_IS_REPLACED ?

No

Reply:
> > I'm not sure I follow. Do you mean eg. shop=butcher should also remove butcher=* ?
> 
> Yes. Looks like I overlooked that. But I also meant generally. Try searching in taginfo for keys named like shop/leisure/craft/... values. I listeed some I found, but I did not look thoroughly.

Well, technically, I could make script do that, but I do not think that is such a good idea. For example, just for `craft=*`, there is `2359` different values according to [taginfo](https://taginfo.openstreetmap.org/keys/craft#values).  `shop=*` is even worse, with `11260` [values](https://taginfo.openstreetmap.org/keys/shop#values). So such heuristics would be adding tens of thousands of keys for removal - vast majority of which does not exist at all in the wild.

The script I've written finds all keys used in combination with `craft=*` and categorizes it if it is used `0.01%` or more times; so if it where listed at least that many times, it would be already categorized (either for removal, or for keeping). (For example for `craft=*`, which is not that popular, that `0.01%` works to only `5` uses worldwide.)  

> Well, right now, **everything** except the specified keys is removed. So when you want to turn this around, you also need to mention every tag that can be removed (i.e. is in that "is kind of shop" expression)... maybe worth a comment in the code, that if the "is kind of shop" expression is extended, your whitelist should be checked too.

I've added a comment to that effect now.

Reply:
> > (it seems wrongly) assumed main tag (the one the quest is about) is removed automatically,
> This might be a good candidate for a test.

Added, since it can't be improved.

Reply:
> This might be a good candidate for a test.

Hm, this is hard though, because *really* the thing we want to test is that our list of keys to remove is exhaustive. So it doesn't particularly help to test a few example tags; the important time to test is when a new tag is added. Which we can't do because the filter expression only returns a boolean.

<details><summary>I did a little dive into how filter expressions work (one more piece of SC demystified :smile:). I'd say it's <i>technically</i> possible to refactor them to return figure out <i>which</i> key matched (allowing automatic removal), but not worth the effort for this minor benefit.</summary>

You'd have to add a new method to each of the [filters](https://github.com/streetcomplete/StreetComplete/tree/master/app/src/main/java/de/westnordost/streetcomplete/data/elementfilter/filters) to return a list of tags that matched. Then change the default implementation of `matches` to check if that list is empty.

However, since I'm pretty sure this code is performance-critical for filtering elements, you'd likely have to keep the old `matches` implementation for many of the filters, so they can short-circuit in some places, so now you've got other functions you need to manually keep in sync— barely an improvement; only worth it if there's some other use for the "which key matched" functionality.
</details>

Reply:
> The script I've written finds all keys used in combination with craft=* and categorizes it if it is used 0.01% or more times

0.01% times of what? Of the particular craft value?

Reply:
> 0.01% times of what? Of the particular craft value?

For `craft` and `shop` for all of them (ie. `craft=*`). For specific ones (ie. `amenity=restaurant`), just for that specific one. 

In short, what taginfo shows as percentage for:
- https://taginfo.openstreetmap.org/keys/craft#combinations
- https://taginfo.openstreetmap.org/tags/amenity=restaurant#combinations 
- etc.

So for example for keys used in combination with `craft` (of which there are `191 204`) I've checked all keys that are used `10` or more times worldwide (should've been `>19` instead of `>10`, but taginfo rounds in different way, so we got more precision). 

Eg. `studio=*`, which is used `26` times worldwide in combination with some `craft=*` is categorized (to be removed), but `certification=*`, which is used `6` times worldwide in combination with some `craft=*` was not categorized.

In **addition** to that, all `fields` and `moreFields` from `id-tagging-schema` (no matter what the percentage)  which you suggested before have been categorized too.

Reply:
> For craft and shop for all of them (ie. craft=*). For specific ones (ie. amenity=restaurant), just for that specific one.

Hm I didn't quite understand that. What I mean is the following (just an example):

If there was a `shop=screws` which is used for at least `x` shops and `screws=*` is tagged more often than `y` of all `shop=screws`, then it should be included in the list.

Looking at taginfo, `x` should be around a magnitude 100, i.e. that seems to be around the place where the values start to lack wiki description. I feel around that number, we can assume that a tag has some reasonable usage. Could be less of course. Some numbers that are not 100 today may be 100 next year. `y` should maybe be the same number. 

Using percentages is problematic because that would mean that with the same absolute number of usages, e.g. `craft=screws` would be included, but `shop=screws` not just because the total number of shops is much higher than the total number of crafts.

Reply:
> Looking at taginfo, x should be around a magnitude 100, i.e. that seems to be around the place where the values start to lack wiki description

I don't think I can accomplish that with taginfo, it lacks resolution. Eg. when I try to see combination tags for `shop=trailer` with `103` worldwide uses, it simply does not display any combination tags at all: https://taginfo.openstreetmap.org/tags/shop=trailer#combinations. 

It seems its cut-off point is at `10 000`, so even `shop=lottery` (with `9 605` uses worldwide) does not list any key [combinations](https://taginfo.openstreetmap.org/tags/shop=lottery#combinations). That is two order of magnitude less precision that you'd like. Do you know of other API able to provide such "combined tags" data with that precision @westnordost ?

Maybe I could just process all combination tags which are used 100 or more times, would that suffice?

Reply:
Yes, you cannot see the combinations for `shop=lottery`, but you can look how many tags there are with the key `lottery`

Reply:
OK @westnordost, I will add that as it seems there is likely no damage that it can cause (apart from that Kotlin variable spanning many more lines). 

However note that my preliminary research show such stats can be very skewed. For example: there is `113` occurrences of `shop=scooter`, so it passes criteria for inclusion; and there is `2 093` occurrences of `scooter=*`, so it passes too. However in reality there is only `11` occurrences of combined `shop=scooter` + `scooter=*` worldwide.

Reply:
@westnordost I've updated script to add such "popular subkeys" (with `x=100`, `y=90`). 
So this should be resolved now.

Reply:
Cool! So, it's ready for merge / review?

Reply:
@westnordost yes, as far as I can tell, there should be no outstanding issues, so it should be ready for review (and merge if it passes review)!